# ðŸ§¬ Multi-Omics Knowledge Graph â€“ TCGA-BRCA

A Knowledge Graph (KG) project for integrating and analyzing **multi-omics data** from the **TCGA-BRCA breast cancer dataset**.  
The graph-based approach connects genomic, transcriptomic, proteomic, and clinical information to support flexible exploration and discovery of biological relationships.

Project report: https://drive.google.com/file/d/1I-CtrcwzPNb3SlWiz6dRBLFwFnRwoEPR/view?usp=drive_link

## Overview

This project models heterogeneous omics data as a graph, where:
- biological and clinical entities â†’ **nodes**
- interactions and associations â†’ **relationships**

The Knowledge Graph enables complex queries and intuitive data exploration for cancer research and precision medicine applications.

## Workflow

1. **Data preprocessing**  
   Extraction and cleaning of key entities (genes, proteins, mutations, pathways, clinical samples)

2. **Graph construction**  
   Modeling and implementation of the KG in **Neo4j**

3. **Querying**  
   Cypher_queries.cypher: Example **Cypher** queries to explore interactions and biological patterns

4. **GDC Graph Analysis**
   GDC_queries.cypher: Centrality (PageRank algorithm) of the KG; community detection (Louvain algorithm); centrality for each community; average centrality inside each community.

## Technologies

- TCGA-BRCA dataset  
- Neo4j with GDC library    
- Python  
- Docker

## Getting started
Run docker compose inside neo4j_container folder to get the Neo4j container
